services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -244,4 +265,101 @@ public BinaryLog getLog(String fullMethodName) {
+   * 8 bytes as the low int64.
+   */
+  // TODO(zpencer): verify int64 representation with other gRPC languages
+  static Uint128 callIdToProto(byte[] bytes) {
+    Preconditions.checkArgument(bytes.length == 16);
+    ByteBuffer bb = ByteBuffer.wrap(bytes);
+    long high = bb.getLong();
+    long low = bb.getLong();
+    return Uint128.newBuilder().setHigh(high).setLow(low).build();
+  }
+
+  @VisibleForTesting
+  // TODO(zpencer): the binlog design does not specify how to actually express the peer bytes
+  static Peer socketToProto(SocketAddress address) {
+    PeerType peerType = null;
+    byte[] peerAddress = null;
+
+    if (address instanceof InetSocketAddress) {
+      InetAddress inetAddress = ((InetSocketAddress) address).getAddress();
+      if (inetAddress instanceof Inet4Address) {
+        peerType = PeerType.PEER_IPV4;
+      } else if (inetAddress instanceof Inet6Address) {
+        peerType = PeerType.PEER_IPV6;
+      } else {
+        logger.log(Level.SEVERE, "unknown type of InetSocketAddress: {}", address);
+      }
+      int port = ((InetSocketAddress) address).getPort();
+      byte[] portBytes = new byte[IP_PORT_BYTES];
+      portBytes[0] = (byte) (port & IP_PORT_UPPER_MASK);
+      portBytes[1] = (byte) (port & IP_PORT_LOWER_MASK);
+      peerAddress = Bytes.concat(inetAddress.getAddress(), portBytes);
+    } else if (address.getClass().getName().equals("io.netty.channel.unix.DomainSocketAddress")) {
+      // To avoid a compile time dependency on grpc-netty, we check against the runtime class name.
+      peerType = PeerType.PEER_UNIX;
+      // DomainSocketAddress.toString() is equivalent to DomainSocketAdddress.path()
+      peerAddress = address.toString().getBytes(Charset.defaultCharset());
+    }
+    Peer.Builder builder = Peer.newBuilder();
+    if (peerType != null) {
+      builder.setPeerType(peerType);
+    }
+    if (peerAddress != null) {
+      builder.setPeer(ByteString.copyFrom(peerAddress));
+    }
+    return builder.build();
+  }
+
+  @VisibleForTesting
+  static io.grpc.binarylog.Metadata metadataToProto(Metadata metadata, int maxHeaderBytes) {
+    Preconditions.checkState(maxHeaderBytes >= 0);
+    Builder builder = io.grpc.binarylog.Metadata.newBuilder();
+    if (maxHeaderBytes > 0) {
+      byte[][] serialized = InternalMetadata.serialize(metadata);
+      int written = 0;
+      // This code is tightly coupled with Metadata's implementation
+      for (int i = 0; i < serialized.length && written < maxHeaderBytes; i += 2) {
+        byte[] key = serialized[i];
+        byte[] value = serialized[i + 1];
+        if (written + key.length + value.length <= maxHeaderBytes) {
+          builder.addEntry(
+              MetadataEntry
+                  .newBuilder()
+                  .setKeyBytes(ByteString.copyFrom(key))
+                  .setValue(ByteString.copyFrom(value))
+                  .build());
+          written += key.length;
```

Comment:
check for overflow?

Reply:
`maxHeaderBytes` should take care of int overflow concerns, though it does not account for message overheads. Is there another overflow you're thinking of?